polygeist/lib/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.cpp

```diff
@@ -0,0 +1,196 @@
+ReachingDefinition::getModifiers(Value val) const {
+  auto it = valueToModifiers.find(val);
+  return (it != valueToModifiers.end()) ? it->second.getArrayRef()
+                                        : std::nullopt;
+}
+
+std::optional<ArrayRef<Operation *>>
+ReachingDefinition::getPotentialModifiers(Value val) const {
+  auto it = valueToPotentialModifiers.find(val);
+  return (it != valueToPotentialModifiers.end()) ? it->second.getArrayRef()
+                                                 : std::nullopt;
+}
+
+//===----------------------------------------------------------------------===//
+// ReachingDefinitionAnalysis
+//===----------------------------------------------------------------------===//
+
+void ReachingDefinitionAnalysis::setToEntryState(ReachingDefinition *lattice) {
+  /// Set the initial state (nothing is known about reaching definitions).
+  propagateIfChanged(lattice, lattice->reset());
+}
+
+void ReachingDefinitionAnalysis::visitOperation(
+    Operation *op, const ReachingDefinition &before,
+    ReachingDefinition *after) {
+  LLVM_DEBUG(llvm::dbgs() << "ReachingDefinitionAnalysis - Visit: " << *op
+                          << "\n");
+
+  // Initialize the alias queries for the current function.
+  if (isa<FunctionOpInterface>(op)) {
+    auto funcOp = cast<FunctionOpInterface>(*op);
+    aliasQueriesMap[funcOp] =
+        std::make_unique<AliasQueries>(funcOp, aliasAnalysis);
+  }
+
+  // If an operation has unknown memory effects assume we can't deduce
+  // anything about reaching definitions.
+  auto memoryEffectOp = dyn_cast<MemoryEffectOpInterface>(op);
+  if (!memoryEffectOp) {
+    LLVM_DEBUG(llvm::dbgs() << "Operation has unknown side effects\n");
+    return setToEntryState(after);
+  }
+
+  // Transfer the input state.
+  ChangeResult result = ChangeResult::NoChange;
+  after->join(before);
+
+  // Retrieve the alias utilities for the function this operation belongs to.
+  auto funcOp = op->getParentOfType<FunctionOpInterface>();
+  AliasQueries &aliasQueries = *aliasQueriesMap[funcOp];
+
+  // Analyze the operation's memory effects.
+  SmallVector<MemoryEffects::EffectInstance> effects;
+  memoryEffectOp.getEffects(effects);
+  for (const auto &effect : effects) {
+    Value val = effect.getValue();
+    if (!val) {
+      // Memory effect on anything other than a value: conservatively assume we
+      // can't deduce anything about reaching definitions.
+      LLVM_DEBUG(llvm::dbgs() << "Memory Effect on non-values found\n");
+      return setToEntryState(after);
+    }
+
+    // Read operations do not modify the reaching definitions state.
+    if (isa<MemoryEffects::Read>(effect.getEffect()))
+      continue;
+
+    // Update the reaching definition for the current value.
+    result |= after->setModifier(val, op);
+
+    // Write operations might also update the reaching definitions of values
+    // that are aliased to the current one.
+    if (isa<MemoryEffects::Write>(effect.getEffect())) {
```

Comment:
what about other memory effects like allocate or free?

Reply:
This is now handled.